Add `prettify-native-form` driver multimethod and refactor `mdb.query/format-sql`

lbrdnk opened the pull request:
Closes #34235

### Description

This PR extends driver with new multimethod, `driver/prettify-native-form`, which is used eg. to pretty format output of `dataset/native` endpoint. Also `mdb.query/format-sql` was refactored to use this method.

Even if we do not currently need to format nosql native code, I've decided to put the method in driver namespace, so it could be used as identity with nosql native code. This way we can avoid branchings or `ignore-exceptions` calls as previously in `/dataset/native` and underlying formatting logic. 

With this method in place, external driver developers can plug in their own formatters or reuse our formatting logic. To find out how, read `driver/prettify-native-form`'s docstring.

### How to verify

Multiple tests were modified during the refactor. Run those and see the results.

### Checklist

- [ ] Tests have been added/updated to cover changes in this PR

lbrdnk committed:
Add multimethod to pretty-format native form

`driver/prettify-native-form` is added. It is added to `driver`
namespace, even though at the time of writing there is no need
to pretty format nosql, eg. mongo, but this way branchings,
as present in `dataset/native`, or rather `mbql.query/format-sql`
can be avoided.

lbrdnk committed:
Move and update `mdb.query/format-sql`

lbrdnk committed:
Add `prettify-native-form` sql implementation

lbrdnk committed:
Refactor endpoint `/dataset/native`

lbrdnk committed:
Update docstring in `prettify-native-form`

`prettify-native-form`has now docstring explaining its relationship
to `sql.u/format-sql`.

lbrdnk committed:
Update tests

lbrdnk committed:
Refactor references in sql-jdbc code

lbrdnk committed:
Implement `prettify-native-form`in drivers

- sqlserver
- oracle
- bigquery-cloud-sdk

lbrdnk committed:
Update test format-sql references

lbrdnk committed:
Refactor modules use of format-sql

lbrdnk committed:
Correct whitespace

metamben commented on docs/developers-guide/driver-changelog.md:
```diff
@@ -70,6 +70,9 @@ title: Driver interface changelog
   removed in 0.51.0 or later. You can easily implement `metabase.driver/db-default-timezone` directly, and use
   `metabase.driver.sql-jdbc.execute/do-with-connection-with-options` to get a `java.sql.Connection` for a Database.
 
+- `metabase.driver/prettify-native-form` was added to enable driver developers use native form formatting
```
I think we should use the word prettify or something similar so as not to confuse things with other formatting like HoneySQL to SQL string.

lbrdnk committed:
Correct whitespace for bigquery

metamben commented on docs/developers-guide/driver-changelog.md:
```diff
@@ -70,6 +70,9 @@ title: Driver interface changelog
   `metabase.driver.sql-jdbc.execute/do-with-connection-with-options` to get a `java.sql.Connection` for a Database.
 
+- `metabase.driver/prettify-native-form` was added to enable driver developers use native form formatting
+  specific to their driver. For details see the issue #34235.
```
I suggest referring to this PR instead of the issue.

metamben commented on modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj:
```diff
@@ -445,3 +446,7 @@
+
+(defmethod driver/prettify-native-form :bigquery-cloud-sdk
+  [_ native-form]
+  (sql.u/format-sql-and-fix-params :mysql native-form))
```
We could call `(driver/prettify-native-form :mysql native-form)`, right? That would require less intimate knowledge of how pretty printing is implemented.

metamben commented on modules/drivers/oracle/src/metabase/driver/oracle.clj:
```diff
@@ -45,6 +45,10 @@
 
+(defmethod driver/prettify-native-form :oracle
+  [_ native-form]
+  (sql.u/format-sql-and-fix-params :plsql native-form))
```
The same. I stop here, and won't repeat this for the other drivers...

metamben commented on src/metabase/db/query.clj:
```diff
@@ -23,50 +23,19 @@
-    sql)))
+  "Return a nicely-formatted version of a `query` string with current application db driver formatting. If you need
+  to use specific driver formatting use [[driver/prettify-native-form]] instead."
+  [sql]
```
Is this `native-query-text` or something similar?

metamben commented on src/metabase/driver.clj:
```diff
@@ -641,6 +641,35 @@
 
+(defmulti prettify-native-form
+  "Pretty-format native form presumably coming from compiled query.
+  Used eg. in API endpoint `/dataset/native`, to present user with nicely formatted query.
```
```suggestion
  Used eg. in the API endpoint `/dataset/native`, to present the user with a nicely formatted query.
```

metamben commented on src/metabase/driver.clj:
```diff
@@ -641,6 +641,35 @@
+    [[metabase.util/ignore-exceptions]] can be avoided as previously needed in `/dataset/native` or former
+    implementation of [[metabase.db.query/format-sql]].
+
+  - What is relationship between `:sql` implementation of this method and [[metabase.driver.sql.util/format-sql]]?
+
+    Function in question is implemented in a way, that driver developers, who would like to use it implementing
+    this multimethod, can:
+    - Avoid implementing it completely, if their driver keyword representation corresponds to key in
+      [[metabase.driver.sql.util/dialects]] (eg. `:postgres`).
+    - Ignore implementing it, if it is sufficient to format their drivers native form with dialect corresponding
+      to `:standardsql`'s value from the dialects map (eg `:h2`).
+    - Use [[metabase.driver.sql.util/format-sql]] in this method's implementation, providing dialect keyword
+      representation that corresponds to to their driver's formatting (eg. `:sqlserver` uses `:tsql`).
+    - Completly reimplement this method with their special formatting code."
```
Can you formulate this as a description how to use and extend this multimethod? There you should explain if and when and how to use `metabase.driver.sql.util/format-sql`. Note that for readers of this code it's probably not very interesting to know that `format-sql` existed before this and I doubt driver implementers used it.

metamben commented on src/metabase/driver/sql.clj:
```diff
@@ -47,6 +48,10 @@
 
+(defmethod driver/prettify-native-form :sql
+  [driver native-form]
+    (sql.u/format-sql-and-fix-params driver native-form))
```
Unexpected indentation.

lbrdnk committed:
Update test

Make `nonsql-dialects-return-original-query-test` run only for mongo
as it now requires initialized mongo driver - mongo paths on class path.

lbrdnk committed:
Update docstring (review suggestions)

For `prettify-native-form`.

lbrdnk committed:
Update docstring (review suggestion)

For `mdb.query/format-sql`.

lbrdnk committed:
Whitespace (review suggestions)

lbrdnk committed:
Update docs (review suggestions)

lbrdnk committed:
Merge branch 'master' into 34235-prettify-native-form
